feat(components/modals): add errors input to modal footer component

Blackbaud-CoreyArcher opened the pull request:
[AB#2460030](https://dev.azure.com/blackbaud/f565481a-7bc9-4083-95d5-4f953da6d499/_workitems/edit/2460030)

Blackbaud-CoreyArcher committed:
feat(components/modals): add errors input to modal footer component

Blackbaud-CoreyArcher commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.scss:
I had the thought that we would apply this style to all status indicators that consumers might already be using. Trevor suggested we scope it to our own implementation. Unsure what the right answer is here

Blackbaud-SteveBrush replied:
I would scope it to our own implementation since we want to avoid altering existing consumer implementations. We can inform consumers of the correct thing to do in our weekly newsletter.

Blackbaud-CoreyArcher commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.ts:
Feedback on docs comment appreciated

blackbaud-sky-build-user commented:
[Storybook preview](https://blackbaud.github.io/skyux-pr-preview/1484/storybook/)

Component Storybooks:

- [action-bars-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/action-bars-storybook/)
- [ag-grid-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/ag-grid-storybook/)
- [avatar-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/avatar-storybook/)
- [data-manager-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/data-manager-storybook/)
- [errors-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/errors-storybook/)
- [flyout-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/flyout-storybook/)
- [forms-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/forms-storybook/)
- [layout-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/layout-storybook/)
- [lookup-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/lookup-storybook/)
- [modals-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/modals-storybook/)
- [pages-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/pages-storybook/)
- [phone-field-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/phone-field-storybook/)
- [split-view-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/split-view-storybook/)
- [tabs-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/tabs-storybook/)
- [text-editor-storybook](https://blackbaud.github.io/skyux-pr-preview/1484/storybooks/text-editor-storybook/)

Apps:

- [code-examples](https://blackbaud.github.io/skyux-pr-preview/1484/code-examples/)
- [dep-graph](https://blackbaud.github.io/skyux-pr-preview/1484/dep-graph/)
- [integration](https://blackbaud.github.io/skyux-pr-preview/1484/integration/)
- [playground](https://blackbaud.github.io/skyux-pr-preview/1484/playground/)

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.scss:
```diff
@@ -16,6 +16,11 @@
+
+  /* Should this be all status indicators or just ours? */
+  .sky-modal-footer-errors .sky-status-indicator {
+    margin-bottom: var(--sky-margin-stacked-lg);
```
Just confirming: Adam recommended this value for the margin?

Blackbaud-CoreyArcher replied:
Yes, Adam and Paul suggested we use this variable

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-SteveBrush commented on apps/playground/src/app/components/modal/modal-error.component.ts:
```diff
@@ -0,0 +1,49 @@
+
+  #onClose(handler: SkyModalBeforeCloseHandler): void {
+    if (this.isError && handler.closeArgs.reason !== 'cancel') {
+      this.errors = ['Sample error', 'Sample error 2'];
```
Can we put a multiline error in here so we can see what it looks like in the playground?

Blackbaud-SteveBrush replied:
I would also create a new modal visual test that includes 2-3 error messages (one modal), with one of the errors being multiline so we can catch any spacing changes in the future.

Blackbaud-CoreyArcher replied:
added

Blackbaud-SteveBrush commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.ts:
```diff
@@ -9,4 +9,11 @@ import { Component, ViewEncapsulation } from '@angular/core';
+export class SkyModalFooterComponent {
+  /**
+   * A collection of error messages to be displayed to the user in the event of a form error.
+   * Values assigned here will be displayed following [SKY UX style guidelines](https://developer.blackbaud.com/skyux/design/guidelines/form-design#validation-and-error-handling).
```
I don't think you need the second sentence since it should be implied that if it's a first-class implementation it will follow SKY UX guidelines. 

The first sentence can be trimmed down a bit (describing "what" instead of "why"): "A list of error messages to be shown above the modal footer buttons."

Blackbaud-CoreyArcher replied:
updated

Blackbaud-SteveBrush reviewed (commented):
I added enough errors to make the modal's height equal to the height of the window and I was unable to "escape" or take action on the modal. We need to account for this rare scenario, especially since it has more likelihood to occur on mobile devices.
<img width="621" alt="Screenshot 2023-07-07 at 9 59 30 AM" src="https://github.com/blackbaud/skyux/assets/12497062/402cd95f-3c8d-469c-b82b-3ca9a1b100f4">

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-CoreyArcher committed:
feedback

Blackbaud-CoreyArcher commented:
> I added enough errors to make the modal's height equal to the height of the window and I was unable to "escape" or take action on the modal. We need to account for this rare scenario, especially since it has more likelihood to occur on mobile devices. <img alt="Screenshot 2023-07-07 at 9 59 30 AM" width="621" src="https://user-images.githubusercontent.com/12497062/251771896-402cd95f-3c8d-469c-b82b-3ca9a1b100f4.png">

After discussing with the team, added some additional styling to cap these messages so they don't interefere with the overall modal layout.

Blackbaud-CoreyArcher committed:
move scrollable container into child element that disappears when there are no errors

Blackbaud-CoreyArcher committed:
move input to parent modal component

Blackbaud-PaulCrowder commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.ts:
```diff
@@ -9,4 +12,14 @@ import { Component, ViewEncapsulation } from '@angular/core';
+  #adapterSvc = inject(SkyModalComponentAdapterService);
+
+  constructor() {
+    this.#adapterSvc.formErrors.subscribe((errors) => {
```
I'd prefer to see the service's `formErrors` property referenced directly in the template using the `async` pipe. That way it will be unsubscribed when the component is destroyed, avoids the "magic" of Angular change detection picking up the change to `formErrors` outside of an input changing, and eliminates a decent amount of code from the class.

Template:

```
  <div aria-live="polite">
    <div *ngIf="(errorsSvc.formErrors | async) as formErrors" class="sky-modal-footer-errors">
      <sky-status-indicator
        *ngFor="let formError of formErrors"
        descriptionType="error"
        indicatorType="danger"
      >
        {{ formError.message }}
      </sky-status-indicator>
    </div>
  </div>
```
Class:

```
  protected errorsSvc = inject(SkyModalErrorsService);
```

Blackbaud-CoreyArcher replied:
updated

Blackbaud-PaulCrowder commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.ts:
```diff
@@ -9,4 +12,14 @@ import { Component, ViewEncapsulation } from '@angular/core';
+export class SkyModalFooterComponent {
+  public formErrors: SkyModalError[] | undefined;
+
+  #adapterSvc = inject(SkyModalComponentAdapterService);
```
The purpose of adapter classes is to isolate DOM manipulation from the rest of the code. Adding a non-DOM-manipulating method or property to it means if we ever want to swap out all the DOM manipulation logic with another service, we'd have to duplicate the non-DOM-manipulation logic. I suggest creating a new `SkyModalErrorsService` with the new property and providing it in `SkyModalComponent`.

Blackbaud-CoreyArcher replied:
added

Blackbaud-PaulCrowder commented on libs/components/modals/src/lib/modules/modal/modal-footer.component.scss:
```diff
@@ -16,6 +16,16 @@
+  .sky-modal-footer-errors {
+    max-height: 100px;
+    overflow-y: auto;
+    margin-bottom: var(--sky-margin-stacked-lg);
```
Instead of adding the margin here, can you remove the rule below and let the last status indicator serve as the bottom margin? Then you can use the `sky-margin-stacked-lg` CSS class on `sky-status-indicator`.

```
.sky-modal-footer-errors {
  max-height: 100px;
  overflow-y: auto;
}
```

In the template: 

```
<sky-status-indicator class="sky-margin-stacked-lg">
```

Blackbaud-CoreyArcher replied:
updated

Blackbaud-CoreyArcher committed:
add errors service, css refactor

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors

Blackbaud-CoreyArcher committed:
Merge branch 'main' into modal-footer-errors
